File: 904-Fruit Into Baskets.py
class Solution(object):
    def totalFruit(self, fruits):
        """
        :type fruits: List[int]
        :rtype: int
        """
        hmap = collections.defaultdict(int)
        N = len(fruits)
        left = 0
        res = 0
        
        for i in range(N):
            hmap[fruits[i]] += 1
            while len(hmap) > 2:
                hmap[fruits[left]] -= 1
                if hmap[fruits[left]] == 0:
                    del hmap[fruits[left]]
                left += 1
            res = max(i-left+1, res)
            
        return res

File: C++/904-Fruit Into Baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        unordered_map<int, int> cnt;
        int res = 0, left = 0;

        for (int i=0; i<fruits.size(); i++) {
            cnt[fruits[i]]++;
            while (cnt.size() > 2) {
                cnt[fruits[left]]--;
                if (cnt[fruits[left]] == 0) {
                    cnt.erase(fruits[left]);
                }
                left++;
            }
            res = max(i-left+1, res);
        }
        return res;
    }
};